hadoop/src/main/java/org/apache/carbondata/hadoop/util/CarbonVectorizedRecordReader.java

```diff
@@ -0,0 +1,194 @@
+    }
+    ++batchIdx;
+    return true;
+  }
+
+
+  private boolean nextBatch() {
+    carbonColumnarBatch.reset();
+    if (iterator.hasNext()) {
+      iterator.processNextBatch(carbonColumnarBatch);
+      numBatched = carbonColumnarBatch.getActualSize();
+      batchIdx = 0;
+      return true;
+    }
+    return false;
+  }
+
+  private void initBatch() {
+    if (carbonColumnarBatch == null) {
+      List<ProjectionDimension> queryDimension = queryModel.getProjectionDimensions();
+      List<ProjectionMeasure> queryMeasures = queryModel.getProjectionMeasures();
+      StructField[] fields = new StructField[queryDimension.size() + queryMeasures.size()];
+      for (ProjectionDimension dim : queryDimension) {
+        fields[dim.getOrdinal()] =
+            new StructField(dim.getColumnName(), dim.getDimension().getDataType());
+      }
+      for (ProjectionMeasure msr : queryMeasures) {
+        DataType dataType = msr.getMeasure().getDataType();
+        if (dataType == DataTypes.BOOLEAN || dataType == DataTypes.SHORT
+            || dataType == DataTypes.INT || dataType == DataTypes.LONG
+            || dataType == DataTypes.FLOAT || dataType == DataTypes.BYTE) {
+          fields[msr.getOrdinal()] =
+              new StructField(msr.getColumnName(), msr.getMeasure().getDataType());
+        } else if (DataTypes.isDecimal(dataType)) {
+          fields[msr.getOrdinal()] = new StructField(msr.getColumnName(),
+              new DecimalType(msr.getMeasure().getPrecision(), msr.getMeasure().getScale()));
+        } else {
+          fields[msr.getOrdinal()] = new StructField(msr.getColumnName(), DataTypes.DOUBLE);
+        }
+      }
+      CarbonColumnVector[] vectors = new CarbonColumnVector[fields.length];
+      for (int i = 0; i < fields.length; i++) {
+        vectors[i] = new CarbonColumnVectorImpl(
+            CarbonV3DataFormatConstants.NUMBER_OF_ROWS_PER_BLOCKLET_COLUMN_PAGE_DEFAULT,
+            fields[i].getDataType());
+      }
+      carbonColumnarBatch = new CarbonColumnarBatch(vectors,
+          CarbonV3DataFormatConstants.NUMBER_OF_ROWS_PER_BLOCKLET_COLUMN_PAGE_DEFAULT,
+          new boolean[] {});
+    }
+  }
+
+  @Override
+  public Object getCurrentValue() throws IOException, InterruptedException {
+    rowCount += 1;
+    Object[] row = new Object[carbonColumnarBatch.columnVectors.length];
+    for (int i = 0; i < carbonColumnarBatch.columnVectors.length; i ++) {
+      if (carbonColumnarBatch.columnVectors[i].getType() == DataTypes.STRING
+          || carbonColumnarBatch.columnVectors[i].getType() == DataTypes.VARCHAR) {
+        byte[] data = (byte[]) carbonColumnarBatch.columnVectors[i].getData(batchIdx - 1);
+        row[i] = ByteUtil.toString(data, 0, data.length);
+      } else {
+        row[i] = carbonColumnarBatch.columnVectors[i].getData(batchIdx - 1);
+      }
+    }
+    return row;
+  }
+
+  @Override public Void getCurrentKey() throws IOException, InterruptedException {
+    return null;
```

Comment:
Don't return null. throw exception if not supported

Reply:
VectorizedCarbonRecordReader is handled in the same way

Reply:
Better add an unsupported exception.